stateful API

nlhepler opened the pull request:
this is a little gross, but it moves tsne into two stages: initialization and optimization. The reason for this is to support an interruptible server that can at least halt optimization soon-ish when it receives a such a request.

Commit:
introduce stateful API

Commit:
add API to free memory, use it

Commit:
safer free_tsne API

Commit:
fix elapsed time reporting

adam-azarchs commented on tsne/bh_sne_src/tsne.cpp:
```diff
@@ -706,16 +714,60 @@ void run(double* X, int N, int D, double* Y, double perplexity, double theta,
-  start = clock();
 
-  for (int iter = 0; iter < max_iter; iter++) {
+  TSNEState* state = new TSNEState;
+  state->P = P;
+  state->row_P = row_P;
+  state->col_P = col_P;
+  state->val_P = val_P;
+  state->dY = dY;
+  state->uY = uY;
+  state->gains = gains;
+  TSNE* tsne = new TSNE;
+  tsne->N = N;
+  tsne->Y = Y;
+  tsne->no_dims = NDIMS;
+  tsne->theta = theta;
+  tsne->max_iter = max_iter;
+  tsne->stop_lying_iter = stop_lying_iter;
+  tsne->mom_switch_iter = mom_switch_iter;
+  tsne->state = state;
+  tsne->iter = 0;
+  tsne->total_time = 0;
+  tsne->residual_time = 0;
```
Use c++11 [aggregate initialization syntax](https://en.cppreference.com/w/cpp/language/aggregate_initialization) for this - it'll be more clear that way and avoid a whole lot of excess pointer indirection.

adam-azarchs commented on tsne/bh_sne_src/tsne.cpp:
```diff
@@ -56,6 +57,16 @@ using std::move;
 using std::sqrt;
 using std::vector;
 
+struct TSNEState {
+  std::vector<double> P;
+  std::vector<unsigned int> row_P;
+  std::vector<unsigned int> col_P;
+  std::vector<double> val_P;
+  std::vector<double> dY;
+  std::vector<double> uY;
+  std::vector<double> gains;
+};
```
Move this inside the anonymous namespace, or move it into the header.  Also, you have `using std::vector` so
```suggestion
struct TSNEState {
  vector<double> P;
  vector<unsigned int> row_P;
  vector<unsigned int> col_P;
  vector<double> val_P;
  vector<double> dY;
  vector<double> uY;
  vector<double> gains;
};
```

adam-azarchs commented on tsne/bh_sne_src/tsne.cpp:
```diff
@@ -576,11 +587,12 @@ double randn() {
-void run(double* X, int N, int D, double* Y, double perplexity, double theta,
-         int rand_seed, bool skip_random_init, double* init, bool use_init,
-         int max_iter, int stop_lying_iter, int mom_switch_iter) {
+struct TSNE* init_tsne(double* X, int N, int D, double* Y, double perplexity,
```
I'd be a lot more comfortable with this using
```suggestion
unique_ptr<TSNEState> init_tsne(double* X, int N, int D, double* Y, double perplexity,
```
Though I do understand that the C api makes that a bit tricky.  It can always use `.get()`, though, thus isolating the badness to inside the `extern "C"` block.

nlhepler replied:
you mean `.release()`, yes?

adam-azarchs commented on tsne/bh_sne_src/tsne.cpp:
```diff
@@ -706,16 +714,60 @@ void run(double* X, int N, int D, double* Y, double perplexity, double theta,
-  start = clock();
 
-  for (int iter = 0; iter < max_iter; iter++) {
+  TSNEState* state = new TSNEState;
```
Certainly here you should be able to use
```suggestion
  unique_ptr<TSNEState> state = make_unique<TSNEState>(
```
with initializer syntax.  Remember there's no types in the C api so the difference between a `unique_ptr` and a `*` pointer doesn't matter there.

nlhepler replied:
added a constructor, ported to make_unique.

adam-azarchs commented on tsne/bh_sne_src/tsne.h:
```diff
@@ -49,12 +49,37 @@
                     int mom_switch_iter = 250);
+// stateful t-SNE
+struct TSNEState;
+struct TSNE {
+  // parameters
+  int N;
+  double* Y;
+  int no_dims;
+  double theta;
+  int max_iter;
+  int stop_lying_iter;
+  int mom_switch_iter;
+  struct TSNEState* state;
+  int iter;
+  float total_time;
+  float residual_time;
+};
```
This should also be a forward-declare, and the actual struct definition should be outside of the `extern "C"` block so that it can use `unique_ptr<TSNEState>` rather than `TSNEState *`.  `TSNEState` is already not a `C` type because it's full of `vector<double>`.  Or even better (and more complex implementation-wise) `vector<double>` for `Y`.  Or just make `state` not be a pointer.  I don't really see why it's a pointer.

Also you should probably reorder a few of the fields for better alignment and packing.  `float` and `int` are 32-bit fields so should be ordered together separately from the pointer fields which have to be 64-bit aligned, and the `double` fields which perform better when 64-bit aligned.

nlhepler replied:
I made `state` a pointer so that it could be forward-declared and `TSNE` could be a C type (`TSNEState` doesn't have to be, it's opaque to C).

adam-azarchs replied:
But as far as the C interface is concerned `TSNE` is an opaque pointer too.  Types are not part of the the C abi, after all.  Where it gets relevant is for cases where you're not just using the ABI, but also interrogating the header file to learn something about the API, e.g. cgo or rust bindgen.  In that case, I think it would be reasonable to stick `state` at the end of the struct and basically telling such consumers to ignore it as something they don't need to pay attention to - lots of structs defined in C APIs have a "`reserved`" region which presumably the internal code understands but is supposed to be opaque to the API consumers.

adam-azarchs replied:
Alternatively, treat `TSNE` as really-opaque as in "you aren't supposed to even know how big it is" and provide accessor methods for the public fields.  Which might be a better idea anyway to ensure users don't violate an invariants.

As far as I can tell the only field on here that are really intended for public consumption is `Y`.  The rest are initialized by `init` and shouldn't be touched by API consumers.

adam-azarchs commented on tsne/bh_sne_src/tsne.cpp:
```diff
@@ -733,68 +785,107 @@ void run(double* X, int N, int D, double* Y, double perplexity, double theta,
+                    double perplexity, double theta, int rand_seed,
+                    bool skip_random_init, double* init, bool use_init,
+                    int max_iter, int stop_lying_iter, int mom_switch_iter) {
+  struct TSNE* tsne = init_tsne(X, N, D, Y, no_dims, perplexity, theta,
+                                rand_seed, skip_random_init, init, use_init,
+                                max_iter, stop_lying_iter, mom_switch_iter);
+  run_n(max_iter, tsne);
+  free_tsne(tsne);
```
This should just lean on `unique_ptr` rather than handling the free manually.

adam-azarchs commented on tsne/bh_sne_src/tsne.cpp:
```diff
@@ -706,16 +714,60 @@ void run(double* X, int N, int D, double* Y, double perplexity, double theta,
+  double theta = tsne->theta;
+  bool exact = (theta == .0) ? true : false;
+  TSNEState* state = tsne->state;
+  vector<double>& dY = state->dY;
+  vector<double>& uY = state->uY;
```
If you're going to do this, you might as well also do it for `P`, `row_P`, `col_P`, and `val_P` to avoid the diffs later on.

nlhepler replied:
Yeah, you're right.

adam-azarchs commented on tsne/bh_sne_src/tsne.cpp:
```diff
@@ -706,16 +714,60 @@ void run(double* X, int N, int D, double* Y, double perplexity, double theta,
+
+// Optimize t-SNE
+template <int NDIMS>
+bool run_n(int n, struct TSNE* tsne) {
```
This really feels like it should be a method on `tsne` rather than a function.  That would let you skip some of the reference-declaring and also would allow the compiler to skip the null-checks it's inserting everywhere.

nlhepler replied:
If I do that, it's not C-api anymore. I could make it all C++-ified, but then I'll also have to craft an independent C-api... which I could do.

adam-azarchs replied:
It's already very, very much not a C api because it's templated.

nlhepler replied:
OK, updated with a C++ API and an `extern "C"` wrapper API for cython, bindgen consumption.

Commit:
fix reviewer comments

this stuffs most pointer badness into extern "C" blocks

adam-azarchs commented on tsne/bh_sne_src/tsne.cpp:
```diff
@@ -54,10 +56,39 @@ using std::fprintf;
+  int iter;
+  float total_time;
+  float residual_time;
+  std::vector<double> P;
+  std::vector<unsigned int> row_P;
+  std::vector<unsigned int> col_P;
+  std::vector<double> val_P;
+  std::vector<double> dY;
+  std::vector<double> uY;
+  std::vector<double> gains;
```
Is all of this really part of the public API?  I'm a bit unconvinced.  Not that it matters really, since the C API is what gets exported and as far as it's concerned this is all an opaque blob.

nlhepler replied:
No, it's not public API. I can and should fix this.

adam-azarchs commented on tsne/bh_sne_src/tsne.cpp:
```diff
@@ -706,9 +735,28 @@ void run(double* X, int N, int D, double* Y, double perplexity, double theta,
-  start = clock();
 
-  for (int iter = 0; iter < max_iter; iter++) {
+  auto tsne = std::unique_ptr<TSNEState>(new TSNEState{
```
Hm, I guess `make_unique` doesn't work unless you wrote a constructor?  The difference is of course purely theoretical so whatever.  However, the more idiomatic way to do this would be
```suggestion
  unique_ptr<TSNEState> tsne(new TSNEState{
```
or perhaps better yet (to make sure there's no question it's an rvalue)
```suggestion
  return unique_ptr<TSNEState>(new TSNEState{
```

nlhepler replied:
Good call.

adam-azarchs commented on tsne/bh_sne_src/tsne.cpp:
```diff
@@ -758,41 +806,92 @@ void run(double* X, int N, int D, double* Y, double perplexity, double theta,
+                               skip_random_init, init, use_init, max_iter,
+                               stop_lying_iter, mom_switch_iter);
+    default:
+      throw "invalid dimension";
```
```suggestion
      throw "unsupported dimension";
```

adam-azarchs commented on tsne/bh_sne_src/tsne.cpp:
```diff
@@ -758,41 +806,92 @@ void run(double* X, int N, int D, double* Y, double perplexity, double theta,
+    case 3:
+      return step_by_impl<3>(step);
+    default:
+      throw "invalid dimension";
```
```suggestion
      throw "unsupported dimension";
```
Though actually come to think of it do we ever actually use 3D?  Doesn't matter a whole lot I guess.

nlhepler replied:
I am strongly in favor of figuring out 3D, it could be a huge draw for our interactive platforms, and people out in the field do find it useful.

adam-azarchs commented on tsne/bh_sne_src/tsne.cpp:
```diff
@@ -758,41 +806,92 @@ void run(double* X, int N, int D, double* Y, double perplexity, double theta,
+  auto tsne = make_tsne(X, N, D, Y, no_dims, perplexity, theta, rand_seed,
+                        skip_random_init, init, use_init, max_iter,
+                        stop_lying_iter, mom_switch_iter);
+  return (TSNE*)tsne.release();
```
I think the newer cast syntax is more clear here in terms of order of operations.
```suggestion
  return static_cast<TSNE*>(tsne.release());
```
At least I _think_ `static_cast` will work there?  Possibly might need `reinterpret_cast`?  Better to use `static_cast` where allowed.

nlhepler replied:
I will figure it out and use the right one.

nlhepler replied:
Have to use `reinterpret_cast` for pointer to pointer casts (except in case of child -> parent class).

adam-azarchs commented on tsne/bh_sne_src/tsne.cpp:
```diff
@@ -758,41 +806,92 @@ void run(double* X, int N, int D, double* Y, double perplexity, double theta,
+void DLL_PUBLIC free_tsne(struct TSNE* tsne) {
+  if (tsne == nullptr)
+    return;
+  delete (TSNEState*)tsne;
```
```suggestion
  delete static_cast<TSNEState*>(tsne);
```

adam-azarchs commented on tsne/bh_sne_src/tsne.cpp:
```diff
@@ -758,41 +806,92 @@ void run(double* X, int N, int D, double* Y, double perplexity, double theta,
+}
+
+bool DLL_PUBLIC step_tsne_by(struct TSNE* tsne, int step) {
+  return ((TSNEState*)tsne)->step_by(step);
```
```suggestion
  return static_cast<TSNEState*>(tsne)->step_by(step);
```

nlhepler replied:
Oh yeah, good call.

adam-azarchs commented on tsne/bh_sne_src/tsne.cpp:
```diff
@@ -758,41 +806,92 @@ void run(double* X, int N, int D, double* Y, double perplexity, double theta,
+  auto tsne = make_tsne(X, N, D, Y, no_dims, perplexity, theta, rand_seed,
+                        skip_random_init, init, use_init, max_iter,
+                        stop_lying_iter, mom_switch_iter);
+  tsne->step_by(max_iter);
```
Nice, much cleaner this way.

adam-azarchs commented on tsne/bh_sne_src/tsne.h:
```diff
@@ -49,12 +49,23 @@
                     int max_iter = 1000, int stop_lying_iter = 250,
                     int mom_switch_iter = 250);
+// stateful t-SNE
+struct TSNE;
```
So the user only actually reads output via `Y` which they passed in?  Cool.  Nice to not have any of those internal state details be part of the C API.

nlhepler replied:
Yeah, once I realized this I just decided to make the whole state object opaque.

adam-azarchs commented on tsne/bh_sne_src/tsne.cpp:
```diff
@@ -758,41 +806,92 @@ void run(double* X, int N, int D, double* Y, double perplexity, double theta,
         fprintf(stderr, "Iteration %d: error is %f\n", iter + 1, C);
       else {
-        total_time += (float)(end - start) / CLOCKS_PER_SEC;
+        elapsed = (float)(end - start) / CLOCKS_PER_SEC;
```
Again, to make order of operations more obvious,
```suggestion
        elapsed = static_cast<float>(end - start) / CLOCKS_PER_SEC;
```
I think there might even be a fixer in clang-tidy or something to do all of those automatically?

adam-azarchs commented on tsne/bh_sne_src/tsne.cpp:
```diff
@@ -54,10 +56,39 @@ using std::fprintf;
+  bool step_by_impl(int n);
+
+ public:
+  int N;
+  double* Y;
+  int no_dims;
+  double theta;
+  int max_iter;
+  int stop_lying_iter;
+  int mom_switch_iter;
```
I think scattering some `const` here and there would be helpful for clarity.
```suggestion
  const int N;
  double const* Y;
  const int no_dims;
  const double theta;
  const int max_iter;
  const int stop_lying_iter;
  const int mom_switch_iter;
```

nlhepler replied:
Y is our output array, but we do want a const memory address, so it should be `double* const Y` no?

adam-azarchs replied:
Yes... it's always a little confusing where the `const` goes with pointers.

adam-azarchs commented on tsne/bh_sne_src/tsne.cpp:
```diff
@@ -706,9 +735,28 @@ void run(double* X, int N, int D, double* Y, double perplexity, double theta,
 
-  for (int iter = 0; iter < max_iter; iter++) {
+  auto tsne = std::unique_ptr<TSNEState>(new TSNEState{
+      N, Y, NDIMS, theta, max_iter, stop_lying_iter, mom_switch_iter, 0, .0, .0,
+      P, row_P, col_P, val_P, dY, uY, gains});
```
Really, `clang-format`?

nlhepler commented:
@adam-azarchs thanks! I really appreciate the detailed review.

Commit:
constify all the things

use reinterpret/static cast and other minor niggles from review

adam-azarchs commented on tsne/bh_sne_src/tsne.cpp:
```diff
@@ -54,10 +56,65 @@ using std::fprintf;
+ public:
+  TSNEState(int N, double* const Y, int no_dims, double theta, int max_iter,
+            int stop_lying_iter, int mom_switch_iter, int iter,
+            float total_time, float residual_time, vector<double> P,
+            vector<unsigned int> row_P, vector<unsigned int> col_P,
+            vector<double> val_P, vector<double> dY, vector<double> uY,
+            vector<double> gains)
```
Unless you're very careful, this is going to end up copying these vectors twice, once when they're passed to the constructor and then again when they're passed to the constructors for the members.  I think you can solve that problem with liberal application of `std::move`.

adam-azarchs replied:
```suggestion
            float total_time, float residual_time, vector<double>&& P,
            vector<unsigned int>&& row_P, vector<unsigned int>&& col_P,
            vector<double>&& val_P, vector<double>&& dY, vector<double>&& uY,
            vector<double>&& gains)
```

Commit:
use move and rvalue refs for TSNEState initializer

adam-azarchs committed:
Update clang-format config to match cellranger.

Main change here is to not bin-pack arguments.

Also add a clang-tidy config.

adam-azarchs committed:
clang-tidy google-readability-braces-around-statements

adam-azarchs committed:
clang-tidy google-readability-casting

adam-azarchs committed:
clang-tidy modernize-use-equals-delete

adam-azarchs committed:
clang-tidy bugprone-reserved-identifier

adam-azarchs committed:
clang-tidy modernize-use-nodiscard

adam-azarchs committed:
clang-tidy bugprone-narrowing-conversions and deadcode.

adam-azarchs committed:
clang-tidy: Adjust types and add constructors.

adam-azarchs committed:
clang-tidy readability-isolate-declaration

adam-azarchs committed:
clang-tidy readability-simplify-boolean-expr

adam-azarchs committed:
clang-tidy readability-static-definition-in-anonymous-namespace

adam-azarchs committed:
clang-tidy readability-inconsistent-declaration-parameter-name

adam-azarchs committed:
clang-tidy readability-redundant-access-specifiers

adam-azarchs committed:
clang-tidy readability-implicit-bool-conversion

adam-azarchs committed:
clang-tidy readability-non-const-parameter

adam-azarchs committed:
Merge #12: Mostly-automatic fixups with clang-tidy.

Commit:
hide GCC pragmas behind ifdef __GNUC__

and visibility fixes for visual studio

Commit:
Merge pull request #13 from 10XDev/lh/vs-fixes

fixes for compilation under Visual Studio

adam-azarchs committed:
Run clang-tidy --fix with clang15.

Also update makefile to use c++17, as that's what we're using on the
rust side.
